src/modules/schulconnex/persistence/schulconnex.repo.ts

```diff
@@ -56,7 +52,49 @@ export class SchulconnexRepo {
                 .where({ 'sp.id': { $in: Array.from(serviceProviderIds) } });
         }
 
-        const results: { personId: string }[] = await qb.orderBy({ 'pk.personId': 'asc' }).execute('all');
-        return Array.from(new Set(results.map((r: { personId: string }) => r.personId))).slice(offset, offset + limit);
+        const results: { personId: string }[] = await qb.execute('all');
+        return Array.from(new Set(results.map((r: { personId: string }) => r.personId)));
+    }
+
+    /**
+     * Finds all unique person IDs that have at least one personenkontext
+     * where the associated rolle has a rollenerweiterung entry for one of the given service providers.
+     * Optionally filters by organisation IDs if provided.
+     *
+     * Supports consistent pagination via stable sorting on `person_id` (if used).
+     *
+     * For performance reasons, a native SQL query is used here to allow a direct join
+     * from the `personenkontext` table to the `rollenerweiterung` table via `rolle_id`,
+     * even though no foreign key exists between these tables.
+     * This avoids unnecessary ORM overhead and enables efficient filtering at the database level.
+     *
+     * @param serviceProviderIds - A set of service provider IDs to filter rollenerweiterung by.
+     * @param organisationIds - (Optional) A set of organisation IDs to filter rollenerweiterung by, or 'all' for no filter.
+     * @returns An array of unique `PersonID`s that match the given criteria.
+     */
+    public async findPersonIdsWithRollenerweiterungForServiceProviderAndOptionallyOrganisations(
+        serviceProviderIds: Set<string>,
+        organisationIds: Set<string> | 'all',
+    ): Promise<PersonID[]> {
+        if (serviceProviderIds.size === 0) {
+            return [];
+        }
+
+        let query: string = `
+            SELECT DISTINCT pk.person_id
+            FROM public.personenkontext pk
+            JOIN public.rollenerweiterung re ON pk.rolle_id = re.rolle_id
```

Comment:
I think we lose the connection of the erweiterung to the organisation here.
```suggestion
            JOIN public.rollenerweiterung re ON pk.rolle_id = re.rolle_id AND pk.organisation_id = re.organisation_id
```

Reply:
Du hast Recht! ist resolved. Danke :)